### PR TITLE
Feature/yolact postproc

### PR DESCRIPTION
**Pull request by qqq1204**
input: net(x), h, w,  top_k,  score_threshold
output: instance_ch, class_ch, dictionary

### CODE REVIEW

File: layers/unloader_postproc.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import cv2
+import numpy as np
+import torch
+
+
+from itertools import chain
+from copy import deepcopy as dp
+from PIL import Image as im
+
+def unloader_pp(det_output,h ,w, top_k = 15, score_threshold = 0.99):
```
Comment (ohilho) [inline layers/unloader_postproc.py:11]: 이 함수가 무얼 하는 함수인지, input이 뭔지 output이 뭔지 하나도 알 수가 없다. 주석을 달아주렴

File: layers/unloader_postproc.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    idx = t[1].argsort(0, descending=True)[:top_k]
+    # Masks are drawn on the GPU, so don't copy    
+    masks = t[3][idx]
+    classes, scores, boxes = [x[idx].cpu().numpy() for x in t[:3]]
+    
+    # masks_np is numpy and for cpu
+    masks_np = (masks).byte().cpu().numpy()
+    
+    """
```
Comment (ohilho) [inline layers/unloader_postproc.py:24]: 이 주석은 무슨 의미니?

File: layers/unloader_postproc.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    classes = np.where(classes == 'your_class', 1,classes)
+    classes = np.where(classes == 'your_class2', 2,classes)
+    """
+
+    instance_ch = np.array(0 ,int)
+    class_ch = np.array(0 ,int)
+    all_new_list =[]
+    tmp_list_for_Gp = []
+    if masks_np.shape[0] == 0:
+      print('imhere')
```
Comment (ohilho) [inline layers/unloader_postproc.py:35]: print가 왜 껴있는지 설명해주렴

**Review by ohilho — COMMENTED** [review]
민재야. 이 코드 돌아가는거 맞는지 난 확신이 안든다. 적어도  PR을 올리기 전에 제대로 작동하는지 확인은 하고 올려야 하는데 확인한거니? 그냥 너가 여태 적어놓은걸 올려두는건 push고 이걸 내가 확인해야 할 단계가 아닌것으로 보인다. 중간에 디버깅하던 코드도 그대로 있는거 같다. 그리고 상당부분 YOLACT의 코드를 복사해서 사용한 것으로 보이는데 어디까지가 복사이고 어디까지가 너가 작성한 것인지 구분해서 주석을 달아두도록 하고, 내가 달아둔 코멘트 전부 reply 달고, 다 했으면 나한테 얘기하렴. 내가 하나하나 YOLACT 코드 다 열어서 한줄씩 비교해가면서 봐줄 시간은 없다.

File: layers/unloader_postproc.py (hunk 1 above)
Comment (ohilho) [reply layers/unloader_postproc.py:11]: [링크](https://github.com/rise-lab-skku/rise-unloader-ros/blob/master/unloader_zmq/scripts/segmentation_server.py) 가 참조가 되면 좋겟다.

Comment (qqq1204) [reply layers/unloader_postproc.py:11]: 형 일단 지금 컴이 윈도우라 내일 아침에 push 할께요. 일단 위에 PR에 설명했던 input은 동일하고 net(x)만 보충 설명 하면 형이 짜신 segmentation_backend.py에 49번째 줄 return이 forward(batch)잖아요. net(x) 자리에 forward(batch)의 return 값을 넣으면 바로 작동하게 넣어놨습니다. 나머지 top_k, score_threshold는 형이 argument로 받아오신걸 쓰면 됩니다.

yolact 에서 가져온건 14번째 post process부터 19번째줄 코드 4줄입니다. 14번째 postprocess 가 이제 행렬곱이랑 크롭해주는 부분입니다. 16, 18, 19번 줄은 저가 짜도 똑같이 나올것같아서 굳이 바꾸지 않았습니다.

File: layers/unloader_postproc.py (hunk 2 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:24]: 이제 masks는 tensor 형태여서 Gpu에 적혀있어서 global 처럼 작동한다는 뜻이고
masks_np 는 numpy 여서 cpu 연산한다는 의미 였습니다.

File: layers/unloader_postproc.py (hunk 3 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:35]: 저가 아까 버그 고치면서 확인할때 넣은것 같아요. 내일 바로 수정해서 push 해 드릴께요.

제 컴퓨터에서는 돌아갔는데 형 컴퓨터에서는 혹시 안돌아가시나요?

File: layers/unloader_postproc.py (hunk 1 above)
Comment (ohilho) [reply layers/unloader_postproc.py:11]: 다른 코드를 설명할 때는 링크를 달아주면 좋겠다. 내가 찾아서 보는 거랑 너가 말하는거랑 동일한지 알 수 없단다. 글고 이 내용을 정리해서 '주석에' 달아주렴. [이렇게](https://github.com/rise-lab-skku/rise-unloader-ros/blob/ee3e193735202afa910e0218020511b5ea9384cd/unloader_zmq/scripts/segmentation_server.py#L86)

File: layers/unloader_postproc.py (hunk 3 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:35]: 맨 마지막 자료형 주석단거도 자세하게 고쳤는데 push가 처음한거로 들어가있네요. 내일 다시 push 할께요 형

File: layers/unloader_postproc.py (hunk 1 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:11]: 아하 알겠습니다. 내일 수정해서 넣을께요

File: layers/unloader_postproc.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    classes, scores, boxes = [x[idx].cpu().numpy() for x in t[:3]]
+    
+    # masks_np is numpy and for cpu
+    masks_np = (masks).byte().cpu().numpy()
+    
+    """
+    class change
+    classes = np.where(classes == 'your_class', 1,classes)
+    classes = np.where(classes == 'your_class2', 2,classes)
+    """
```
Comment (ohilho) [inline layers/unloader_postproc.py:28]: 이 '주석을' 말한 거였다. 이건 무슨 의미니?

File: layers/unloader_postproc.py (hunk 3 above)
Comment (ohilho) [reply layers/unloader_postproc.py:35]: 그래 pr 하기 전에는 확인을 잘 하고 올리자.

**Review by ohilho — COMMENTED** [review]
첫 pr이라 당연히 완벽할거라 생각 안했으니 걱정말고 하나씩 해보자. 전체적으로 무슨 코드인지 pr에서 구구절절 설명해야 한다면 잘못된 pr이라고 생각한다. pr에서는 무엇을 잘했고 무엇을 고치면 좋을지 피드백을 받도록 하고 전체를 다 보면서 하나하나 설명하는 자리가 되지 않도록 하자. 뒷부분은 앞부분이 설명이 안되니까 추측하면서 읽고 있었는데 결국 추측을 한 다는것 자체가 pr에 의미가 없다고 생각이 들어서 중단하고 다시 작성하자고 한거야.

File: layers/unloader_postproc.py (hunk 3 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:35]: 네 형 내일 아침에 올리겠습니다.

File: layers/unloader_postproc.py (hunk 4 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:28]: 아 저거 혹시 클래스가 int 1, 2, 3, 4 가 아닐경우 저런식으로 고쳐서 써라. 라는 의미였어요

**Review by ohilho — COMMENTED** [review]
코멘트 추가했다.

File: layers/unloader_postproc.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    instance_ch = np.array(0 ,int)
+    class_ch = np.array(0 ,int)
+    all_new_list =[]
+    tmp_list_for_Gp = []
```
[the remark this answers is not in the document]
Comment (qqq1204) [reply layers/unloader_postproc.py:32]: Grasping point 약자로 적었습니다.

File: layers/unloader_postproc.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        all_new_list.append(tmp_dict.copy())
+
+    all_new_list = sorted(all_new_list, key=lambda k: k['size'], reverse = True) 
+
+    # intergrate images
+    for i in range(len(classes)):
+        #class channel 
+        class_ch1 = np.array(0 ,int)
+        tmp_class = int(all_new_list[i]['classes'])
+        class_ch1 = all_new_list[i]['masks'] * tmp_class
```
[the remark this answers is not in the document]
Comment (qqq1204) [reply layers/unloader_postproc.py:60]: int는 immutable한 객체여서 deepcopy를 안해도 되는데, immutable하면 할당 할 때마다 메모리 주소가 변경되서 그렇습니다.

File: layers/unloader_postproc.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    dict_for_Gp.setdefault('seg_length',[])
+    dict_for_Gp.setdefault('scores',[])
+    dict_for_Gp.setdefault('bboxes',[])
+    
+    # make new list for sort by size of mask and set index of all pixels
+    for i in range(len(classes)):
+        tmp_dict = {}
+        tmp_dict = {'size': masks_np[i, :, :].sum(),'masks': masks_np[i, :, :], 'classes': classes[i],
+       'scores': scores[i], 'bboxes': boxes[i]}
+        all_new_list.append(tmp_dict.copy())
```
[the remark this answers is not in the document]
Comment (qqq1204) [reply layers/unloader_postproc.py:51]: 형 말하신거 따라서 고쳤습니다.

File: layers/unloader_postproc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (qqq1204) [reply layers/unloader_postproc.py:32]: 후안할때 보낼 메세지 이름이랑 동일하게 Gp 부분만 YolactSegm 로 고쳤습니다.

File: layers/unloader_postproc.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        class_ch [numpy.adarray]: class channel with image size
+        dict_for_Gp[dictionary]: {'segm_masks':[np.int32_list],'seg_length': [int_list], 
+                                'scores' : [np.float32_list], 'bboxes': [np.int64_list], 
+                                'image_size' : (int,int) <- (h, w) , 'num_objs': int}
+                                
+                                segm_masks: contour x,y pixel
+                                seg_length: the number of contour pixel point
+                                bboxes: 4 pixel coordinate for bounding boxes
+                                num_objs: the number of detected
+    """
```
Comment (ohilho) [inline layers/unloader_postproc.py:33]: 전보다 훨씬 알아보기 좋게 잘 썻다.  몇 가지 궁금한 것은 return 에서 타입을 명시하는데 bracket ('[', ']')을 사용했는데 python에서는 이게 리스트라는 의미라서 주석이 불명확해진다. 예를 들어 "seg_length " key에 할당된 value가 List(int) 인것인지 List(List(int)) 인 것인지 헷갈린다. 타입을 표기할 때 numpy.ndarray 같은건 타입 이름이 명확하지만 list와 dict의 표현에는 주의가 필요하다. python의 typing module의 표기를 따라서 List(int)나 기본 파이썬 문법을 따라서 [int] 라고 적어주렴. int_list라는 타입은 선언한 적이 없어보인다. 그리고 numpy.adarray라고 오타가 있는거 같다.

File: layers/unloader_postproc.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    t = postprocess(det_output, w, h, score_threshold = score_threshold)
+     # see https://github.com/dbolya/yolact/blob/master/eval.py #L149
```
Comment (ohilho) [inline layers/unloader_postproc.py:36]: 주석이 위에 오도록.

File: layers/unloader_postproc.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    #initial var
+    instance_ch = np.array(0 ,int)
+    class_ch = np.array(0 ,int)
+    all_new_list =[]
+    tmp_list_for_Gp = []
+
+    #none instance case
+    if masks_np.shape[0] == 0:
+        tmp_list_for_Gp= {'seg_masks':[],'num_objs': 0, 'image_size': (h,w),
+                     'seg_length':[], 'scores': [], 'bboxes': []}
```
Comment (ohilho) [inline layers/unloader_postproc.py:55]: tmp_list_for_GP 는 50번째 줄에서 list로 초기화 했는데 여기서는 dictionary이다. 
이름과 자료형이 매치가 안된다. 초기화도 필요 없어보인다.

File: layers/unloader_postproc.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ohilho) [reply layers/unloader_postproc.py:69]: ? 수정이 안되어있는데?

File: layers/unloader_postproc.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ohilho) [reply layers/unloader_postproc.py:78]: 그렇다면 all_new_list[i]['masks'] * tmp_class 의 결과가 무슨 타입인지 확인해보렴. 내 생각엔 Iterable일거 같은데 결과의 타입을 확인해보고 76번 줄과의 관계를 생각해본 후 나에게 어떻게 하는게 좋을지 말해주렴

File: layers/unloader_postproc.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        dict_for_Gp['bboxes'].append(all_new_list[i]['bboxes'])
+        '''
+        Check the contour result
+
+        img3 = cv2.drawContours(cond3, [contours[0]], -1, (255,0,0), 3)
+        img3 = im.fromarray(img3*50)
+        img3.save('output_images/output_contour'+str(i)+'.png')
+        '''
+    #flatten and save list of sementation and boxes
+    flatten_list = [j for sub in tmp_list_for_Gp for j in sub]
```
Comment (ohilho) [inline layers/unloader_postproc.py:116]: list of list를 flatten 하는 방법으로 generator expression을 사용하는게 느릴 것이라 생각이 드는데 더 좋은 방법이 없나 찾아보자.

File: layers/unloader_postproc.py (hunk 9 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:36]: commit 하고 push를 하기 전에 수정하면 push를 2번 해야하나 봅니다. 바뀐게 적용이 안되어있네요.

File: layers/unloader_postproc.py (hunk 10 above)
Comment (qqq1204) [reply layers/unloader_postproc.py:55]: 네임 바꾸겠습니다. 여기는 초기화보다 이제 사물이 0개 일때 이렇게 따로 자료형만 같은 null값들을 넣어주지 않으면 코드가 여기저기 돌아갈 때 오류를 일으킬까봐 넣어놨습니다.  바꿀까요?

File: layers/unloader_postproc.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        instance_ch (numpy.ndarray): instance channel with image size
+        class_ch (numpy.ndarray): class channel with image size
+        dict_for_YolactSegm (dictionary): {'segm_masks':List(np.int32),'seg_length': List(int), 
+                                'scores' : List(np.float32), 'bboxes': List(np.int64), 
+                                'image_size' : List(int) <- (h, w) , 'num_objs': int}
+                                
+                                segm_masks: contour x,y pixel
+                                seg_length: the number of contour pixel point
+                                bboxes: 2 corners coordinate for bounding boxes
+                                num_objs: the number of detected
```
Comment (ohilho) [inline layers/unloader_postproc.py:32]: scores 와 image_size 에 대한 타입만 위에 적고 내용은 아래에 따로 적을 것.

File: layers/unloader_postproc.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        dict_for_YolactSegm['bboxes'].append(all_new_list[i]['bboxes'])
+        '''
+        Check the contour result
+
+        img3 = cv2.drawContours(cond3, [contours[0]], -1, (255,0,0), 3)
+        img3 = im.fromarray(img3*50)
+        img3.save('output_images/output_contour'+str(i)+'.png')
+        '''
+    #flatten and save list of sementation and boxes
+    flatten_list = list(chain.from_iterable(tmp_list_for_YolactSegm))
```
Comment (ohilho) [inline layers/unloader_postproc.py:120]: 굿.

File: layers/unloader_postproc.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
+from .output_utils import postprocess
+import cv2
+import numpy as np
+import torch
+
```
Comment (ohilho) [inline layers/unloader_postproc.py:5]: 빈칸 정리

**Review by ohilho — COMMENTED** [review]
처음에 비해 많이 좋다. 마무리 해보자
